### PR TITLE
Remove as many generic error() messages from base as possible

### PR DESCRIPTION
**Pull request by jakebolewski**
- throw more specific Exception types
- make error messages more consistent
- give more context for the error when possible
- update tests

### CODE REVIEW

File: base/socket.jl [hunk 1]
```diff
@@ -21,9 +21,9 @@ end
 
 function IPv4(host::Integer)
     if host < 0
-        error("IP address must not be negative")
+        throw(ArgumentError("IPv4 address must not be ≥ 0"))
```
Comment (GlenHertz) [inline base/socket.jl]: Isn't the sense on this statement wrong?  The IPv4 must be positive.

File: base/socket.jl [hunk 2]
```diff
@@ -63,11 +63,11 @@ end
 
 function IPv6(host::Integer)
     if host < 0
-        error("IP address must not be negative")
+        throw(ArgumentError("IPv6 address must not be ≥ 0"))
```
Comment (GlenHertz) [inline base/socket.jl]: Wrong sense here too?

**Comment by ViralBShah** [comment]
This is great, and much needed.

File: base/socket.jl (hunk 1 above)
Comment (IainNZ) [inline base/socket.jl]: Do these inequality symbols display correctly on Windows?

Comment (jakebolewski) [inline base/socket.jl]: @GlenHertz thanks for reading through.  "must be positive" I think is subjectively better than "not negative", I'll push an update.

@IainNZ I have no idea.

**Comment by jakebolewski** [comment]
I've updated with @GlenHertz's suggestions and did another pass over what I pushed last night.  I also updated some of the errors in the `LinearAlgebra` module.

**Comment by IainNZ** [comment]
Really nice work @jakebolewski !

**Comment by staticfloat** [comment]
:+1:

**Pull request merged**